File: backend/src/lib/routes/topics.py
from fastapi import APIRouter, Depends, HTTPException
from typing import List
from pydantic import BaseModel
from backend.src.lib.auth.middleware import require_auth
from backend.src.lib.auth.service import get_current_user
from backend.src.lib.db.models.topic import TopicModel
from shared.src.types import Topic, User, LessonPlan

router = APIRouter(prefix="/topics", tags=["topics"])

class CreateTopicRequest(BaseModel):
    title: str
    description: str
    lesson_plan: LessonPlan

class UpdateTopicRequest(BaseModel):
    title: str
    description: str
    lesson_plan: LessonPlan

@router.get("/", response_model=List[Topic])
async def get_topics(current_user: User = Depends(require_auth())):
    """
    Get all topics for the current user.
    If user has admin role, returns all topics.
    """
    try:
        # Get topics for the user
        topics = await TopicModel.getByUserId(current_user.id)
        return topics
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch topics: {str(e)}"
        )

@router.post("/", response_model=Topic)
async def create_topic(
    topic_data: CreateTopicRequest,
    current_user: User = Depends(require_auth())
):
    """
    Create a new topic.
    Only users with admin role can create topics.
    """
    try:
        # Check if user has admin role
        if "role_admin" not in current_user.roles:
            raise HTTPException(
                status_code=403,
                detail="Only administrators can create topics"
            )

        # Create the topic
        topic = await TopicModel.create(
            current_user.id,
            topic_data.title,
            topic_data.description,
            topic_data.lesson_plan
        )
        return topic
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to create topic: {str(e)}"
        )

@router.get("/{topic_id}", response_model=Topic)
async def get_topic(
    topic_id: str,
    current_user: User = Depends(require_auth())
):
    """
    Get a specific topic by ID.
    Users can only access their own topics unless they have admin role.
    """
    try:
        topic = await TopicModel.getById(topic_id)
        if not topic:
            raise HTTPException(
                status_code=404,
                detail="Topic not found"
            )

        # Check if user has access to this topic
        if topic.userId != current_user.id and "role_admin" not in current_user.roles:
            raise HTTPException(
                status_code=403,
                detail="You don't have permission to access this topic"
            )

        return topic
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch topic: {str(e)}"
        )

@router.put("/{topic_id}", response_model=Topic)
async def update_topic(
    topic_id: str,
    topic_data: UpdateTopicRequest,
    current_user: User = Depends(require_auth())
):
    """
    Update a specific topic.
    Only admins or the topic owner can update it.
    """
    try:
        # Check if topic exists and user has access
        existing_topic = await TopicModel.getById(topic_id)
        if not existing_topic:
            raise HTTPException(
                status_code=404,
                detail="Topic not found"
            )

        # Check permissions
        if existing_topic.userId != current_user.id and "role_admin" not in current_user.roles:
            raise HTTPException(
                status_code=403,
                detail="You don't have permission to update this topic"
            )

        # Update the topic
        updated_topic = await TopicModel.update(
            topic_id,
            topic_data.title,
            topic_data.description,
            topic_data.lesson_plan
        )
        return updated_topic
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to update topic: {str(e)}"
        )

@router.delete("/{topic_id}")
async def delete_topic(
    topic_id: str,
    current_user: User = Depends(require_auth())
):
    """
    Delete a specific topic.
    Only admins or the topic owner can delete it.
    """
    try:
        # Check if topic exists and user has access
        existing_topic = await TopicModel.getById(topic_id)
        if not existing_topic:
            raise HTTPException(
                status_code=404,
                detail="Topic not found"
            )

        # Check permissions
        if existing_topic.userId != current_user.id and "role_admin" not in current_user.roles:
            raise HTTPException(
                status_code=403,
                detail="You don't have permission to delete this topic"
            )

        # Delete the topic
        await TopicModel.delete(topic_id)
        return {"message": "Topic deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to delete topic: {str(e)}"
        )

File: backend/src/routes/topics.ts
import express from 'express';
import { TopicService } from '../lib/services/topic';
import { authenticateToken } from '../middleware/auth';

export const topicsRouter = express.Router();

// Get all topics
topicsRouter.get('/', authenticateToken, async (req, res) => {
  try {
    const topics = await TopicService.getAllTopics();
    res.json(topics);
  } catch (error) {
    console.error('Error getting topics:', error);
    res.status(500).json({ error: 'Failed to get topics' });
  }
});

// Get a specific topic
topicsRouter.get('/:topicId', authenticateToken, async (req, res) => {
  try {
    const topic = await TopicService.getTopic(Number(req.params.topicId));
    if (!topic) {
      return res.status(404).json({ error: 'Topic not found' });
    }
    res.json(topic);
  } catch (error) {
    console.error('Error getting topic:', error);
    res.status(500).json({ error: 'Failed to get topic' });
  }
});

// Create a new topic
topicsRouter.post('/', authenticateToken, async (req, res) => {
  try {
    const { title, description, difficulty } = req.body;
    if (!title) {
      return res.status(400).json({ error: 'Title is required' });
    }
    const topic = await TopicService.createTopic({
      title,
      description,
      difficulty: difficulty || 'beginner',
      user_id: req.user!.user_id
    });
    res.status(201).json(topic);
  } catch (error) {
    console.error('Error creating topic:', error);
    res.status(500).json({ error: 'Failed to create topic' });
  }
});

// Update a topic
topicsRouter.patch('/:topicId', authenticateToken, async (req, res) => {
  try {
    const { title, description, difficulty } = req.body;
    const topic = await TopicService.updateTopic(Number(req.params.topicId), {
      title,
      description,
      difficulty,
      user_id: req.user!.user_id
    });
    if (!topic) {
      return res.status(404).json({ error: 'Topic not found' });
    }
    res.json(topic);
  } catch (error) {
    console.error('Error updating topic:', error);
    res.status(500).json({ error: 'Failed to update topic' });
  }
});

// User-specific routes
// Start a topic for a user
topicsRouter.post('/me/:topicId/start', authenticateToken, async (req, res) => {
  try {
    const { goalText } = req.body;
    const targetDate = req.body.targetDate ? new Date(req.body.targetDate) : undefined;
    
    if (!goalText) {
      return res.status(400).json({ error: 'Goal text is required' });
    }

    const userTopic = await TopicService.startUserTopic(
      req.user!.user_id,
      Number(req.params.topicId),
      goalText,
      targetDate
    );
    res.status(201).json(userTopic);
  } catch (error) {
    console.error('Error starting topic:', error);
    res.status(500).json({ error: 'Failed to start topic' });
  }
});

// Get user's topic progress
topicsRouter.get('/me/:topicId/progress', authenticateToken, async (req, res) => {
  try {
    const progress = await TopicService.getTopicProgress(
      req.user!.user_id,
      Number(req.params.topicId)
    );
    res.json(progress);
  } catch (error) {
    console.error('Error getting topic progress:', error);
    res.status(500).json({ error: 'Failed to get topic progress' });
  }
});

// Get user's topic details
topicsRouter.get('/me/:topicId', authenticateToken, async (req, res) => {
  try {
    const userTopic = await TopicService.getUserTopic(
      req.user!.user_id,
      Number(req.params.topicId)
    );
    if (!userTopic) {
      return res.status(404).json({ error: 'User topic not found' });
    }
    res.json(userTopic);
  } catch (error) {
    console.error('Error getting user topic:', error);
    res.status(500).json({ error: 'Failed to get user topic' });
  }
});

// Update user's topic
topicsRouter.patch('/me/:topicId', authenticateToken, async (req, res) => {
  try {
    const { currentLessonId, goalText, targetDate } = req.body;
    const userTopic = await TopicService.updateUserTopic(
      req.user!.user_id,
      Number(req.params.topicId),
      {
        currentLessonId: currentLessonId ? Number(currentLessonId) : undefined,
        goalText,
        targetDate: targetDate ? new Date(targetDate) : undefined
      }
    );
    if (!userTopic) {
      return res.status(404).json({ error: 'User topic not found' });
    }
    res.json(userTopic);
  } catch (error) {
    console.error('Error updating user topic:', error);
    res.status(500).json({ error: 'Failed to update user topic' });
  }
});
